src/lua/luaTask.c

```diff
@@ -124,117 +146,206 @@ size_t get_ontick_freq()
+
+                lua_pop(g_lua,1);
+                return false;
+        }
 
-    xTaskCreate( luaTask,
-                 ( signed portCHAR * ) "luaTask",
-                 LUA_STACK_SIZE,
-                 NULL,
-                 priority,
-                 NULL);
+        pr_info("SUCCESS!\r\n");
+        /* Empty the stack? --> lua_settop (g_lua, 0); */
+
+        lua_gc(g_lua, LUA_GCCOLLECT,0);
+        pr_info("lua: memory usage: ");
+        pr_info_int(lua_gc(g_lua, LUA_GCCOUNT,0));
+        pr_info("Kb\r\n");
+
+        return true;
 }
 
-static void doScript(void)
+void initialize_lua()
 {
-    lockLua();
-    const char *script = getScript();
-    size_t len = strlen(script);
-    pr_info("running lua script len(");
-    pr_info_int(len);
-    pr_info(")...");
+        lockLua();
 
-    lua_gc(g_lua, LUA_GCCOLLECT,0);
+        pr_info("lua: Starting...\r\n");
 
-    int result = (luaL_loadbuffer(g_lua, script, len, "startup") || lua_pcall(g_lua, 0, LUA_MULTRET, 0));
-    if (0 != result) {
-        pr_error("lua: startup script error: (");
-        pr_error(lua_tostring(g_lua,-1));
-        pr_error(")\r\n");
-        lua_pop(g_lua,1);
-    }
-    pr_info("done\r\n");
-    lua_gc(g_lua, LUA_GCCOLLECT,0);
-    pr_info("lua: memory usage: ");
-    pr_info_int(lua_gc(g_lua, LUA_GCCOUNT,0));
-    pr_info("Kb\r\n");
-    unlockLua();
+        if (LUA_ENABLED == lua_run_state)
+                goto out;
+
+        g_lua = lua_newstate(myAlloc, NULL);
+        if (!g_lua) {
+                pr_error("LUA: Can't allocate memory for LUA state.\r\n");
+                goto out;
+        }
+
+        //open optional libraries
+        luaopen_base(g_lua);
+        luaopen_table(g_lua);
+        luaopen_string(g_lua);
+        luaopen_math(g_lua);
+        registerBaseLuaFunctions(g_lua);
+        registerLuaLoggerBindings(g_lua);
+
+        if (!_load_script()) {
+                _terminate_lua();
+                goto out;
+        }
+
+        /* If here, then init was successful.  Enable runtime */
+        lua_run_state = LUA_ENABLED;
+
+out:
+        unlockLua();
 }
 
-static void guardedDoScript(void)
+static void run_lua_method(const char *method)
 {
-    if (!watchdog_is_watchdog_reset()) {
-        doScript();
-    } else {
-        pr_error("lua: Crash detected- bypassing Lua script for recovery mode\r\n");
-        LED_enable(3);
-    }
+        lockLua();
+
+        lua_getglobal(g_lua, method);
+        if (lua_isnil(g_lua, -1)) {
+                pr_error("Can't find \"");
+                pr_error(method);
+                pr_error("\"\r\n");
+
+                lua_pop(g_lua, 1);
+                _terminate_lua();
+                goto out;
+        }
+
+        if (0 != lua_pcall(g_lua, 0, 0, 0)) {
+                pr_error("lua method error: ");
+                pr_error(lua_tostring(g_lua, -1));
+                pr_error("\r\n");
+
+                lua_pop(g_lua, 1);
+                _terminate_lua();
```

Comment:
> I'm noting the aggressive termination of Lua on errors seems to be out of scope of the stated intent of the PR.

Correct.  I got a little zealous here :).  I understand the concerns RE killing the script of a pcall fails.  I can pull out the agressive killing.